cranelift-codegen/src/simple_preopt.rs

```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
     }
 }
 
+struct BranchOptInfo {
+    br_inst: Inst,
+    cmp_arg: Value,
+    args: ValueList,
+    new_opcode: Opcode,
+}
+
+/// Fold comparisons into branch operations when possible.
+///
+/// This matches against operations which compare against zero, then use the
+/// result in a `brz` or `brnz` branch. It folds those two operations into a
+/// single `brz` or `brnz`.
+fn branch_opt(pos: &mut FuncCursor, inst: Inst) {
+    let mut info = if let InstructionData::Branch {
+        opcode, ref args, ..
+    } = pos.func.dfg[inst]
+    {
+        let first_arg = {
+            let args = pos.func.dfg.inst_args(inst);
+            args[0]
+        };
+
+        let iconst_inst =
+            if let ValueDef::Result(iconst_inst, _) = pos.func.dfg.value_def(first_arg) {
+                iconst_inst
+            } else {
+                return;
+            };
+
+        if let InstructionData::IntCompareImm {
+            opcode: Opcode::IcmpImm,
+            arg: cmp_arg,
+            cond: cmp_cond,
+            imm: cmp_imm,
+        } = pos.func.dfg[iconst_inst]
+        {
+            let cmp_imm: i64 = cmp_imm.into();
+            if cmp_imm != 0 {
+                return;
+            }
+
+            // icmp_imm returns non-zero when the comparison is true. So, if
+            // we're branching on zero, we need to invert the condition.
+            let cond = match opcode {
+                Opcode::Brz => cmp_cond.inverse(),
+                Opcode::Brnz => cmp_cond,
+                _ => return,
+            };
+
+            let new_opcode = match cond {
+                IntCC::Equal => Opcode::Brz,
+                IntCC::NotEqual => Opcode::Brnz,
+                _ => return,
+            };
+
+            BranchOptInfo {
+                br_inst: inst,
+                cmp_arg: cmp_arg,
+                args: args.clone(),
+                new_opcode: new_opcode,
+            }
+        } else {
+            return;
+        }
+    } else {
+        return;
+    };
+
+    info.args.as_mut_slice(&mut pos.func.dfg.value_lists)[0] = info.cmp_arg;
+    if let InstructionData::Branch { ref mut opcode, .. } = pos.func.dfg[info.br_inst] {
+        *opcode = info.new_opcode;
+    } else {
+        panic!();
+    }
+}
+
+struct BranchOrderInfo {
```

Comment:
Since this struct is only used internally within a function, could we use a tuple in the function instead, and that would remove the need for this structure?

Reply:
I went ahead and did this. But I'm really not sure that it improved readability of the function. Take a look and let me know what you think?

Reply:
Definitely an improvement *after* the let statement. It's maybe less clear within the body, but it seems there are few enough lines in between that it's easy to check the tuple definition, so I'd be good with keeping it the new way. Thanks!